### PR TITLE
[fix][client] Cannot access message data inside ProducerInterceptor#onSendAcknowledgement

### PR DESCRIPTION
**Pull request by Shawyeok**
### Motivation

In #22393, the payload of message get released before calling `ProducerInterceptor#onSendAcknowledgement`, hence access message data inside `ProducerInterceptor#onSendAcknowledgement` will get exception below:
```
io.netty.util.IllegalReferenceCountException: refCnt: 0
	at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1454) ~[netty-buffer-4.1.115.Final.jar:4.1.115.Final]
	at io.netty.buffer.UnpooledHeapByteBuf.array(UnpooledHeapByteBuf.java:147) ~[netty-buffer-4.1.115.Final.jar:4.1.115.Final]
	at org.apache.pulsar.client.impl.MessageImpl.getData(MessageImpl.java:371) ~[classes/:?]
	at org.apache.pulsar.client.api.InterceptorsTest$6.onSendAcknowledgement(InterceptorsTest.java:239) ~[test-classes/:?]
	at org.apache.pulsar.client.api.interceptor.ProducerInterceptorWrapper.onSendAcknowledgement(ProducerInterceptorWrapper.java:58) ~[classes/:?]
	at org.apache.pulsar.client.impl.ProducerInterceptors.onSendAcknowledgement(ProducerInterceptors.java:100) [classes/:?]
	at org.apache.pulsar.client.impl.ProducerBase.onSendAcknowledgement(ProducerBase.java:162) [classes/:?]
	at org.apache.pulsar.client.impl.ProducerImpl$DefaultSendMessageCallback.onSendComplete(ProducerImpl.java:451) [classes/:?]
	at org.apache.pulsar.client.impl.ProducerImpl$DefaultSendMessageCallback.sendComplete(ProducerImpl.java:423) [classes/:?]
	at org.apache.pulsar.client.impl.ProducerImpl$OpSendMsg.sendComplete(ProducerImpl.java:1643) [classes/:?]
	at org.apache.pulsar.client.impl.ProducerImpl.ackReceived(ProducerImpl.java:1316) [classes/:?]
	at org.apache.pulsar.client.impl.ClientCnx.handleSendReceipt(ClientCnx.java:490) [classes/:?]
	at org.apache.pulsar.common.protocol.PulsarDecoder.channelRead(PulsarDecoder.java:236) [classes/:?]
```

### Modifications

- Release the message payload after `ProducerInterceptor#onSendAcknowledgement` is invoked to ensure the payload is accessible during the callback
- Add test case `InterceptorsTest#testProducerInterceptorAccessMessageData`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - *InterceptorsTest#testProducerInterceptorAccessMessageData*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/Shawyeok/pulsar/pull/22

### CODE REVIEW

**Review by poorbarcode — APPROVED** [review]
LGTM

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 1]
```diff
@@ -436,20 +436,22 @@ private void onSendComplete(Throwable e, SendCallback sendCallback, MessageImpl<
...
+                    sendCallback.getFuture().completeExceptionally(e);
+                } else {
+                    latencyHistogram.recordSuccess(latencyNanos);
+                    publishedBytesCounter.add(msgSize);
+                    stats.incrementNumAcksReceived(latencyNanos);
+                    onSendAcknowledgement(msg, msg.getMessageId(), null);
+                    sendCallback.getFuture().complete(msg.getMessageId());
+                }
+            } finally {
+                ReferenceCountUtil.safeRelease(payload);
```
Comment (thetumbled) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:454]: Could we remove the null check for payload?

**Review by dao-jun — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 1 above)
Comment (Shawyeok) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:454]: I assume you are talking the code below:
```java
            if (payload == null) {
                log.error("[{}] [{}] Payload is null when calling onSendComplete, which is not expected.",
                        topic, producerName);
            }
```

This seems harmless, I have no clue about original intention here, so let's keep this patch simple and stupid :->

Comment (thetumbled) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:454]: This patch change the original logic.

```
if (payload == null) {
                log.error("[{}] [{}] Payload is null when calling onSendComplete, which is not expected.",
                        topic, producerName);
            } else {
                ReferenceCountUtil.safeRelease(payload);
            }
```
Before this patch, only payload is not null we will release it, while after this patch, the null check is removed.

Comment (Shawyeok) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:454]: Oh, `ReferenceCountUtil.safeRelease` is null-safe as its name suggests. It does nothing if invoked with a `null` value, and this behavior is explicitly documented in its Javadoc. 

https://github.com/netty/netty/blob/c7de17a7f385a8a2156ddc9e74c7a85a6651a034/common/src/main/java/io/netty/util/ReferenceCountUtil.java#L114

**Review by thetumbled — APPROVED** [review]
LGTM.

**Pull request merged**